File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_Android/JavaWrapperDemo/wrapper/src/main/java/com/hyphenate/wrapper/EMPushManagerWrapper.java
package com.hyphenate.wrapper;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMPushConfigs;
import com.hyphenate.chat.EMPushManager;
import com.hyphenate.chat.EMSilentModeParam;
import com.hyphenate.chat.EMSilentModeResult;
import com.hyphenate.exceptions.HyphenateException;

import com.hyphenate.wrapper.util.EMHelper;
import com.hyphenate.wrapper.util.EMSDKMethod;
import com.hyphenate.wrapper.callback.EMCommonCallback;
import com.hyphenate.wrapper.callback.EMCommonValueCallback;
import com.hyphenate.wrapper.callback.EMWrapperCallback;
import com.hyphenate.wrapper.helper.EMConversationHelper;
import com.hyphenate.wrapper.helper.EMPushConfigsHelper;
import com.hyphenate.wrapper.helper.EMSilentModeParamHelper;
import com.hyphenate.wrapper.helper.EMSilentModeResultHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class EMPushManagerWrapper extends EMBaseWrapper{
    EMPushManagerWrapper() {
        registerEaseListener();
    }

    public String onMethodCall(String method, JSONObject jsonObject, EMWrapperCallback callback) throws JSONException {
        String ret = null;
        if (EMSDKMethod.getImPushConfig.equals(method)) {
            ret = getImPushConfig(jsonObject, callback);
        }
        else if(EMSDKMethod.getImPushConfigFromServer.equals(method)){
            ret = getImPushConfigFromServer(jsonObject, callback);
        }
        else if(EMSDKMethod.updatePushNickname.equals(method)){
            ret = updatePushNickname(jsonObject, callback);
        }
        else if(EMSDKMethod.updateImPushStyle.equals(method)){
            ret = updateImPushStyle(jsonObject, callback);
        }
        else if(EMSDKMethod.updateHMSPushToken.equals(method)){
            ret = updateHMSPushToken(jsonObject, callback);
        }
        else if(EMSDKMethod.updateFCMPushToken.equals(method)){
            ret = updateFCMPushToken(jsonObject, callback);
        }
        else if (EMSDKMethod.reportPushAction.equals(method)) {
            ret = reportPushAction(jsonObject, callback);
        }
        else if (EMSDKMethod.setConversationSilentMode.equals(method)) {
            ret = setConversationSilentMode(jsonObject, callback);
        }
        else if (EMSDKMethod.removeConversationSilentMode.equals(method)) {
            ret = removeConversationSilentMode(jsonObject, callback);
        }
        else if (EMSDKMethod.fetchConversationSilentMode.equals(method)) {
            ret = fetchConversationSilentMode(jsonObject, callback);
        }
        else if (EMSDKMethod.setSilentModeForAll.equals(method)) {
            ret = setSilentModeForAll(jsonObject, callback);
        }
        else if (EMSDKMethod.fetchSilentModeForAll.equals(method)) {
            ret = fetchSilentModeForAll(jsonObject, callback);
        }
        else if (EMSDKMethod.fetchSilentModeForConversations.equals(method)) {
            ret = fetchSilentModeForConversations(jsonObject, callback);
        }
        else if (EMSDKMethod.setPreferredNotificationLanguage.equals(method)) {
            ret = setPreferredNotificationLanguage(jsonObject, callback);
        }
        else if (EMSDKMethod.fetchPreferredNotificationLanguage.equals(method)) {
            ret = fetchPreferredNotificationLanguage(jsonObject, callback);
        }
        else if (EMSDKMethod.getPushTemplate.equals(method)) {
            ret = getPushTemplate(jsonObject, callback);
        }
        else if (EMSDKMethod.setPushTemplate.equals(method)) {
            ret = setPushTemplate(jsonObject, callback);
        } else {
            ret = super.onMethodCall(method, jsonObject, callback);
        }
        return ret;
    }

    private String getImPushConfig(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigs();
        return EMHelper.getReturnJsonObject(EMPushConfigsHelper.toJson(configs)).toString();
    }

    private String getImPushConfigFromServer(JSONObject params, EMWrapperCallback callback) throws JSONException {
        asyncRunnable(()->{
            try {
                EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigsFromServer();
                JSONObject jo = null;
                try {
                    JSONObject jsonObject = EMPushConfigsHelper.toJson(configs);
                    jo = jsonObject;
                }catch (JSONException e) {
                    e.printStackTrace();
                }finally {
                    onSuccess(jo, callback);
                }
            } catch (HyphenateException e) {
                onError(e, callback);
            }
        });
        return null;
    }

    private String updatePushNickname(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String nickname = params.getString("nickname");

        asyncRunnable(()->{
            try {
                EMClient.getInstance().pushManager().updatePushNickname(nickname);
                onSuccess(nickname, callback);
            } catch (HyphenateException e) {
                onError(e, callback);
            }
        });
        return null;
    }

    private String updateImPushStyle(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMPushManager.DisplayStyle style = params.getInt("pushStyle") == 0 ? EMPushManager.DisplayStyle.SimpleBanner : EMPushManager.DisplayStyle.MessageSummary;
        EMClient.getInstance().pushManager().asyncUpdatePushDisplayStyle(style, new EMCommonCallback(callback));
        return null;
    }

    private String updateHMSPushToken(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String token = params.getString("token");
        asyncRunnable(()->{
            EMClient.getInstance().sendHMSPushTokenToServer(token);
            onSuccess(token, callback);
        });
        return null;
    }

    private String updateFCMPushToken(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String token = params.getString("token");
        String fcmKey = EMClient.getInstance().getOptions().getPushConfig().getFcmSenderId();
        EMClient.getInstance().pushManager().bindDeviceToken(fcmKey, token, new EMCommonCallback(callback));
        return null;
    }

    private String reportPushAction(JSONObject params, EMWrapperCallback callback) throws JSONException {
        return null;
    }

    private String setConversationSilentMode(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String conversationId = params.getString("convId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("conversationType"));
        EMSilentModeParam param = EMSilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForConversation(conversationId, type, param, new EMCommonValueCallback<EMSilentModeResult>(callback){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                JSONObject jo = null;
                try {
                    jo = EMSilentModeResultHelper.toJson(object);
                } catch (JSONException e) {
                    e.printStackTrace();
                } finally {
                    updateObject(jo);
                }
            }
        });
        return null;
    }
    private String removeConversationSilentMode(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("convType"));
        EMClient.getInstance().pushManager().clearRemindTypeForConversation(conversationId, type, new EMCommonCallback(callback));
        return null;
    }

    private String fetchConversationSilentMode(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("convType"));
        EMClient.getInstance().pushManager().getSilentModeForConversation(conversationId, type, new EMCommonValueCallback<EMSilentModeResult>(callback){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                JSONObject jo = null;
                try {
                    jo = EMSilentModeResultHelper.toJson(object);
                } catch (JSONException e) {
                    e.printStackTrace();
                } finally {
                    updateObject(jo);
                }
            }
        });
        return null;
    }

    private String setSilentModeForAll(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMSilentModeParam param =  EMSilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForAll(param ,new EMCommonValueCallback<EMSilentModeResult>(callback){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                JSONObject jo = null;
                try {
                    jo = EMSilentModeResultHelper.toJson(object);
                } catch (JSONException e) {
                    e.printStackTrace();
                } finally {
                    updateObject(jo);
                }
            }
        });
        return null;
    }

    private String fetchSilentModeForAll(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMClient.getInstance().pushManager().getSilentModeForAll(new EMCommonValueCallback<EMSilentModeResult>(callback){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                JSONObject jo = null;
                try {
                    jo = EMSilentModeResultHelper.toJson(object);
                } catch (JSONException e) {
                    e.printStackTrace();
                } finally {
                    updateObject(jo);
                }
            }
        });
        return null;
    }
    private String fetchSilentModeForConversations(JSONObject params, EMWrapperCallback callback) throws JSONException {
        Iterator iterator = params.keys();
        ArrayList<EMConversation> list = new ArrayList<>();
        while (iterator.hasNext()) {
            String conversationId = (String)iterator.next();
            EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt(conversationId));
            EMConversation conversation = EMClient.getInstance().chatManager().getConversation(conversationId, type, true);
            list.add(conversation);
        }

        EMClient.getInstance().pushManager().getSilentModeForConversations(list, new EMCommonValueCallback<Map<String, EMSilentModeResult>>(callback) {
            @Override
            public void onSuccess(Map<String, EMSilentModeResult> object) {
                JSONObject result = new JSONObject();
                try {
                    for (Map.Entry<String, EMSilentModeResult>entry: object.entrySet()) {
                        result.put(entry.getKey(), EMSilentModeResultHelper.toJson(entry.getValue()));
                    }
                }catch (JSONException e) {
                    e.printStackTrace();
                }finally {
                    updateObject(result);
                }
            }
        });
        return null;
    }

    private String setPreferredNotificationLanguage(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String code = params.getString("code");
        EMClient.getInstance().pushManager().setPreferredNotificationLanguage(code, new EMCommonCallback(callback));
        return null;
    }

    private String fetchPreferredNotificationLanguage(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMClient.getInstance().pushManager().getPreferredNotificationLanguage(new EMCommonValueCallback<>(callback));
        return null;
    }

    private String setPushTemplate(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String pushTemplateName = params.getString("pushTemplateName");
        EMClient.getInstance().pushManager().setPushTemplate(pushTemplateName, new EMCommonCallback(callback));
        return null;
    }

    private String getPushTemplate(JSONObject params, EMWrapperCallback callback) throws JSONException {
        EMClient.getInstance().pushManager().getPushTemplate(new EMCommonValueCallback<>(callback));
        return null;
    }
    
    private void registerEaseListener(){}
}

File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_iOS/wrapper/wrapper/managers/EMPushManagerWrapper.m
//
//  EMPushManagerWrapper.m
//  wrapper
//
//  Created by 杜洁鹏 on 2022/11/15.
//

#import "EMPushManagerWrapper.h"
#import "EMPushOptions+Helper.h"
#import "EMConversation+Helper.h"
#import "EMGroup+Helper.h"
#import "EMSilentModeParam+Helper.h"
#import "EMSilentModeResult+Helper.h"
#import "EMHelper.h"
#import "EMUtil.h"

@implementation EMPushManagerWrapper
- (instancetype)init {
    if (self = [super init]) {
        [self registerEaseListener];
    }
    
    return self;
}

- (NSString *)onMethodCall:(NSString *)method
                    params:(NSDictionary *)params
                  callback:(EMWrapperCallback *)callback {
    NSString *ret = nil;
    if ([getImPushConfig isEqualToString:method]) {
        ret = [self getImPushConfig:params callback:callback];
    }
    else if([getImPushConfigFromServer isEqualToString:method]){
        ret = [self getImPushConfigFromServer:params callback:callback];
    }
    else if([updatePushNickname isEqualToString:method]){
        ret = [self updatePushNickname:params callback:callback];
    }
    else if([updateImPushStyle isEqualToString:method]){
        ret = [self updateImPushStyle:params callback:callback];
    }
    else if ([reportPushAction isEqualToString:method]) {
        ret = [self reportPushAction:params callback:callback];
    }
    else if ([setConversationSilentMode isEqualToString:method]) {
        ret = [self setConversationSilentMode:params callback:callback];
    }
    else if ([removeConversationSilentMode isEqualToString:method]) {
        ret = [self removeConversationSilentMode:params callback:callback];
    }
    else if ([fetchConversationSilentMode isEqualToString:method]) {
        ret = [self fetchConversationSilentMode:params callback:callback];
    }
    else if ([setSilentModeForAll isEqualToString:method]) {
        ret = [self setSilentModeForAll:params callback:callback];
    }
    else if ([fetchSilentModeForAll isEqualToString:method]) {
        ret = [self fetchSilentModeForAll:params callback:callback];
    }
    else if ([fetchSilentModeForConversations isEqualToString:method]) {
        ret = [self fetchSilentModeForConversations:params callback:callback];
    }
    else if ([setPreferredNotificationLanguage isEqualToString:method]) {
        ret = [self setPreferredNotificationLanguage:params callback:callback];
    }
    else if ([fetchPreferredNotificationLanguage isEqualToString:method]) {
        ret = [self fetchPreferredNotificationLanguage:params callback:callback];
    }
    else if ([getPushTemplate isEqualToString:method]) {
        ret = [self getPushTemplate:params callback:callback];
    }
    else if ([setPushTemplate isEqualToString:method]) {
        ret = [self setPushTemplate:params callback:callback];
    }else {
        ret = [super onMethodCall:method params:params callback:callback];
    }
    return ret;
}


- (NSString *)getImPushConfig:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    EMPushOptions *options = EMClient.sharedClient.pushManager.pushOptions;
    return [[EMHelper getReturnJsonObject:[options toJson]] toJsonString];
}

- (NSString *)getImPushConfigFromServer:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    [EMClient.sharedClient.pushManager getPushNotificationOptionsFromServerWithCompletion:^(EMPushOptions *aOptions, EMError *aError)
     {
        [weakSelf wrapperCallback:callback error:aError object:[aOptions toJson]];
    }];
    return nil;
}

- (NSString *)updatePushNickname:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *nickname = params[@"nickname"];
    [EMClient.sharedClient.pushManager updatePushDisplayName:nickname
                                                  completion:^(NSString * _Nonnull aDisplayName, EMError * _Nonnull aError)
     {
        [weakSelf wrapperCallback:callback error:aError object:aDisplayName];
    }];
    return nil;
}

- (NSString *)updateImPushStyle:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    EMPushDisplayStyle pushStyle = [params[@"pushStyle"] intValue];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *aError = [EMClient.sharedClient.pushManager updatePushDisplayStyle:pushStyle];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallback:callback error:aError object:nil];
        });
    });
    return nil;
}

- (NSString *)reportPushAction:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    // TODO： no support
    return nil;
}

- (NSString *)setConversationSilentMode:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *conversaionId = params[@"convId"];
    EMConversationType type = [EMConversation typeFromInt:[params[@"convType"] intValue]];
    EMSilentModeParam *silmentParam = [EMSilentModeParam formJson:params[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                             params:silmentParam
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallback:callback error:aError object:[aResult toJson]];
    }];
    return nil;
}

- (NSString *)removeConversationSilentMode:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *conversaionId = params[@"convId"];
    EMConversationType type = [EMConversation typeFromInt:[params[@"convType"] intValue]];
    [EMClient.sharedClient.pushManager clearRemindTypeForConversation:conversaionId
                                                     conversationType:type
                                                           completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallback:callback error:aError object:[aResult toJson]];
    }];
    return nil;
}

- (NSString *)fetchConversationSilentMode:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *conversaionId = params[@"convId"];
    EMConversationType type = [EMConversation typeFromInt:[params[@"convType"] intValue]];
    [EMClient.sharedClient.pushManager getSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:[aResult toJson]];
    }];
    return nil;
}

- (NSString *)setSilentModeForAll:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    EMSilentModeParam *silmentParam = [EMSilentModeParam formJson:params[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForAll:silmentParam
                                                completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:[aResult toJson]];
    }];
    return nil;
}

- (NSString *)fetchSilentModeForAll:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    [EMClient.sharedClient.pushManager getSilentModeForAllWithCompletion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallback:callback error:aError object:[aResult toJson]];
    }];
    return nil;
}

- (NSString *)fetchSilentModeForConversations:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSMutableArray *conversations = [NSMutableArray array];
    for (NSString *conversaitonId in params.allKeys) {
        EMConversationType type = [EMConversation typeFromInt:[params[conversaitonId] intValue]];
        EMConversation *conversation = [EMClient.sharedClient.chatManager getConversation:conversaitonId type:type createIfNotExist:YES];
        [conversations addObject:conversation];
    }
    
    
    [EMClient.sharedClient.pushManager getSilentModeForConversations:conversations
                                                          completion:^(NSDictionary<NSString *,EMSilentModeResult *> * _Nullable aResult, EMError * _Nullable aError)
     {
        
        NSMutableDictionary *tmpDict = [NSMutableDictionary dictionary];
        for (NSString *conversationId in aResult.allKeys) {
            tmpDict[conversationId] = [aResult[conversationId] toJson];
        }
        
        [weakSelf wrapperCallback:callback error:aError object:tmpDict];
    }];
    return nil;
}

- (NSString *)setPreferredNotificationLanguage:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *code = params[@"code"];
    [EMClient.sharedClient.pushManager setPreferredNotificationLanguage:code completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:nil];
    }];
    return nil;
}

- (NSString *)fetchPreferredNotificationLanguage:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    [EMClient.sharedClient.pushManager getPreferredNotificationLanguageCompletion:^(NSString * _Nullable aLaguangeCode, EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:aLaguangeCode];
    }];
    return nil;
}

- (NSString *)getPushTemplate:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    NSString *pushTemplateName = params[@"pushTemplateName"];
    [EMClient.sharedClient.pushManager setPushTemplate:pushTemplateName completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:nil];
    }];
    return nil;
}

- (NSString *)setPushTemplate:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    __weak EMPushManagerWrapper *weakSelf = self;
    [EMClient.sharedClient.pushManager getPushTemplate:^(NSString * _Nullable aPushTemplateName, EMError * _Nullable aError) {
        [weakSelf wrapperCallback:callback error:aError object:aPushTemplateName];
    }];
    return nil;
}

- (void)registerEaseListener{
    
}

@end
